Source/hydro/trace_plm.cpp

```diff
@@ -0,0 +1,297 @@
+#include "Castro.H"
+#include "Castro_F.H"
+#include "Castro_hydro_F.H"
+
+#ifdef RADIATION
+#include "Radiation.H"
+#endif
+
+#include <cmath>
+
+#include <ppm.H>
+
+using namespace amrex;
+
+void
+Castro::trace_plm(const Box& bx, const int idir,
+                  Array4<Real const> const q_arr,
+                  Array4<Real const> const qaux_arr,
+                  Array4<Real const> const dq,
+                  Array4<Real> const qm,
+                  Array4<Real> const qp,
+#if AMREX_SPACEDIM < 3
+                  Array4<Real const> const dloga,
+#endif
+                  Array4<Real const> const srcQ,
+                  const Box& vbx,
+                  const Real dt) {
+
+  // here, bx is the box we loop over -- this can include ghost cells
+  // vbx is the valid box (no ghost cells)
+
+  const auto dx = geom.CellSizeArray();
+
+  const Real dtdx = dt/dx[idir];
+
+  auto lo = bx.loVect3d();
+  auto hi = bx.hiVect3d();
+
+  auto vlo = vbx.loVect3d();
+  auto vhi = vbx.hiVect3d();
+
+#ifndef AMREX_USE_CUDA
+  if (ppm_type != 0) {
+    std::cout << "Oops -- shouldnt be in tracexy with ppm_type != 0" << std::endl;
+    amrex::Error("Error:: trace_3d.f90 :: tracexy");
+  }
+#endif
+
+  int QUN;
+  int QUT;
+  int QUTT;
+
+  if (idir == 0) {
+    QUN = QU;
+    QUT = QV;
+    QUTT = QW;
+
+  } else if (idir == 1) {
+    QUN = QV;
+    QUT = QW;
+    QUTT = QU;
+
+  } else {
+    QUN = QW;
+    QUT = QU;
+    QUTT = QV;
+  }
+
+  Real lsmall_dens = small_dens;
+  Real lsmall_pres = small_pres;
+
+  GpuArray<int, npassive> qpass_map_p;
```

Comment:
I recommend using the new `upassmap` and `qpassmap` inline functions rather than this array capture.

Reply:
done